miden-crypto/src/dsa/ecdsa_secp256k1/key_agreement.rs

```diff
@@ -0,0 +1,121 @@
+//! scenario. Namely, when the sender knows the (static) public key of the receiver and it
+//! uses that, together with an ephemeral secret key that it generates, to derive a shared
+//! secret.
+//! This shared secret will then be used to encrypt some message (using for example a key
+//! derivation function).
+//! The public key associated with the ephemeral secret key will be sent alongside the encrypted
+//! message.
+
+use hkdf::{Hkdf, hmac::SimpleHmac};
+use k256::{
+    AffinePoint,
+    elliptic_curve::rand_core::{CryptoRng, RngCore},
+    sha2::Sha256,
+};
+
+use super::PublicKey;
+
+/// A shared secret computed using the ECDH (Elliptic Curve Diffie-Hellman) key agreement.
+pub struct SharedSecret {
+    inner: k256::ecdh::SharedSecret,
+}
+
+impl SharedSecret {
+    /// Returns a HKDF (HMAC-based Extract-and-Expand Key Derivation Function) that can be used
+    /// to extract entropy from the shared secret.
+    /// This basically converts a shared secret into uniformly random values that are appropriate
+    /// for use as key material.
+    pub fn extract<D>(&self, salt: Option<&[u8]>) -> Hkdf<Sha256, SimpleHmac<Sha256>> {
+        self.inner.extract(salt)
+    }
+
+    pub(crate) fn from_inner(inner: k256::ecdh::SharedSecret) -> SharedSecret {
+        Self { inner }
+    }
+}
+
+/// Ephemeral public key for ECDH key agreement over secp256k1 curve.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct EphemeralPublicKey {
+    inner: k256::PublicKey,
+}
+
+impl EphemeralPublicKey {
+    /// Returns a reference to this ephemeral public key as an elliptic curve point in affine
+    /// coordinates.
+    pub fn as_affine(&self) -> &AffinePoint {
+        self.inner.as_affine()
+    }
+}
+
+/// Ephemeral secret key for ECDH key agreement over secp256k1 curve.
+pub struct EphemeralSecretKey {
+    inner: k256::ecdh::EphemeralSecret,
+}
+
+impl EphemeralSecretKey {
+    /// Generates a new random ephemeral secret key using the OS random number generator.
+    #[cfg(feature = "std")]
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        let mut rng = k256::elliptic_curve::rand_core::OsRng;
+
+        Self::with_rng(&mut rng)
+    }
+
+    /// Generates a new ephemeral secret key using the provided random number generator.
+    pub fn with_rng<R: CryptoRng + RngCore>(rng: &mut R) -> Self {
+        let sk_e = k256::ecdh::EphemeralSecret::random(rng);
+        Self { inner: sk_e }
+    }
+
+    /// Gets the corresponding ephemeral public key for this ephemeral secret key.
+    pub fn ephemeral_public_key(&self) -> EphemeralPublicKey {
+        let pk = self.inner.public_key();
+        EphemeralPublicKey { inner: pk }
+    }
+
+    /// Computes a Diffie-Hellman shared secret from an ephemeral secret key and the (static) public
+    /// key of the other party.
+    pub fn diffie_hellman(&self, pk_other: PublicKey) -> SharedSecret {
```

Comment:
Should this take `EphemeralSecretKey` instead of `PublicKey` as a parameter?

Reply:
The method is basically on a secret key (more precisely `EphemeralSecretKey`) and so the parameter needs to be a public key. In the optimal case, this would be an ephemeral pubic key, but for our use case, where we want to avoid interaction, this is just the normal (static) public key.

The usage of this method will be as follows:

1. The sender will create an ephemeral secret key
2. The sender will use the (static) public key of the receiver in the above method to derive a shared secret
3. The sender will use shared secret to encrypt and will attach the ephemeral public key (corresponding to the ephemeral secret key) to the encrypted note to be sent through the network